### PR TITLE
Reconnect on node loss

### PR DESCRIPTION
**Pull request by George-Payne**
- wrap calls in execute to capture errors
- check for errors that require reconnection
- wrap streaming calls with `createGRPCStream`
- listen to errors and check if reconnection is required
- reconnect on connection errors

fixes: #176 

RFC: https://github.com/EventStore/home/issues/574

### CODE REVIEW

File: src/Client/index.ts [hunk 1]
```diff
@@ -320,6 +362,52 @@ export class Client {
     return this.#channel;
   };
 
+  private shouldReconnect = (
+    err: Error
+  ): [shouldReconnect: boolean, to?: EndPoint] => {
```
Comment (oskardudycz) [inline src/Client/index.ts:367]: **TIL:** I didn't know this syntax 😮

File: src/__test__/connection/reconnect.test.ts [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        await delay(200);
+      }
+    }
+
+    expect(i).toBeLessThan(20);
+
+    await cluster.down();
+  });
+});
+
```
Comment (oskardudycz) [inline src/__test__/connection/reconnect.test.ts:173]: **SUGGESTION:** It might be worth verifying what will happen if all the nodes are down (or not accessible) and they get back or when a node is killed during reconnection (that may be probably hard to simulate). It might be also worth making sure that if we get an error that should not cause reconnection then no reconnection is made.

It would be also good to verify somewhere if the order of preferences is applied properly.

**Review by oskardudycz — APPROVED** [review]
It looks good, but it would be good if we could address the comment about the test coverage: https://github.com/EventStore/EventStore-Client-NodeJS/pull/177/files#r675339897.

**Review by YoEight — APPROVED** [review]

File: src/__test__/connection/reconnect.test.ts (hunk 2 above)
Comment (George-Payne) [reply src/__test__/connection/reconnect.test.ts:244]: - [ ]  if all the nodes are down (or not accessible) and they get back
- [ ]  a node is killed during reconnection
- [ ]  if we get an error that should not cause reconnection then no reconnection is made

File: src/__test__/utils/Cluster.ts [hunk 3]
```diff
@@ -295,11 +293,65 @@ export class Cluster {
...
   private path = (...parts: string[]): string => {
     return join(__dirname, "instances", this.id, ...parts);
   };
 
   private cleanUp = async () => {
-    await rmdir(this.path(), { recursive: true });
+    try {
+      await rmdir(this.path(), { recursive: true });
+    } catch (error) {
+      // TODO:
```
Comment (oskardudycz) [inline src/__test__/utils/Cluster.ts:341]: **QUESTION:** Will this be done in a separate PR?

File: src/__test__/utils/Cluster.ts [hunk 4]
```diff
@@ -295,11 +293,65 @@ export class Cluster {
...
+
+        try {
+          const response = await exec(
+            node,
+            `curl --fail --insecure http${
+              this.insecure ? "" : "s"
+            }://localhost:2113/health/live`,
+            { cwd: this.path() }
+          );
+          // console.log(
```
Comment (oskardudycz) [inline src/__test__/utils/Cluster.ts:315]: **QUESTION:** Do we want to leave this commented out `console.log` (and others in this file)?

**Review by oskardudycz — APPROVED** [review]

**Review by thefringeninja — APPROVED** [review]
